Add external streams to assets

jbblanchet opened the pull request:
First draft of adding external streams to assets. Fixes #75 and #30. I updated an undocumented piece of code that appeared to do a similar job. If this is an acceptable solution, I'll create tests and update the doc.

Sample usage (with typescript):

```javascript
// create stream of virtual files
var tsStream = gulp.src("src/**/*.ts")
        .pipe(ts());

    // pass stream to assets, multiple streams are supported
    var assets = useref.assets({}, tsStream);

    // use gulp-useref normally
    return gulp.src("src/index.html")
        .pipe(assets)
        .pipe(assets.restore())
        .pipe(useref())
        .pipe(gulp.dest("dist"));
```

jbblanchet commented:
This appears to change the code created by  #108, but it is undocumented, untested and I couldn't make it work.

jonkemp commented:
I'm unclear on what this does. Does it add support for #75 and #30 or does it fix #70?

jbblanchet commented:
I misunderstood the intention of #70, and I now see that it is unrelated to the passing of assets streams. I've updated the code accordingly.

Usage is now:
```javascript
// create stream of virtual files
var tsStream = gulp.src("src/**/*.ts")
        .pipe(ts());

    // pass stream to assets, multiple streams are supported
    var assets = useref.assets({ additionalStreams: tsStream });

    // use gulp-useref normally
    return gulp.src("src/index.html")
        .pipe(assets)
        .pipe(assets.restore())
        .pipe(useref())
        .pipe(gulp.dest("dist"));
```

This pull request is not ready to be merged. File order is not preserved between physical and virtual files, and I have an issue with using gulp-inject in the same workflow that I'm still looking into. This is more a request for comments or suggestions. :)

jbblanchet commented:
Ok, this takes care of the ordering and the race condition with gulp-inject. Example workflow is:

```javascript
var tsStream = gulp.src("src/**/*.ts")
        .pipe(ts());

    var assets = useref.assets({ additionalStreams: tsStream });

    return gulp.src("src/index.html")
        .pipe(inject(tsStream, {relative: true}))
        .pipe(assets)
        .pipe(assets.restore())
        .pipe(useref())
        .pipe(gulp.dest("dist"));
```

I'll test it with my full workflow (wiredep, rev, etc) on monday when I'm back at the office, but for now it looks good. I'd like to know if you'd be willing to pull, and if so, I'll update the docs and create some automated tests.

jonkemp commented:
I'm interested. I need to look at it more in depth though before I can decide. 
Just to be clear, the feature requests in #75 and #30 are what this is addressing?
Thanks!

jbblanchet commented:
Yes, this is addressing both these issues. I'm using typescript here, but I could also be using sass, less, coffeescript, dart, angular-templatecache, etc.

Currently if I want to use gulp-useref with typescript, I have to generate the files, save them to a temp directory, then use gulp-useref. This pull request lets me pass streams of generated files to useref instead, which is more "gulpy".

In my last example I'm using gulp-inject to automatically add the script tags for the files generated by typescript to my index.html before running gulp-useref.

Let me know if you have any questions.

Thanks.

jonkemp commented:
Yes, looks good. Please continue this work. Thanks!

jbblanchet commented:
Ok, I think I finally solved the race condition between the streams. Changes appear worst than they really are since I add to re-indent most of the assets function. Basically, the change since the last version is waiting for the additional assets to run the main assets function. Still trying to break it, will let you know when I'm done.

jonkemp commented:
Was this done or forgotten?

jbblanchet commented:
Neither. The code looks good to go, I was planning on creating some automated tests and updating the readme this weekend.

jbblanchet commented:
All done.

jonkemp commented:
Can you squash your commits?

jbblanchet commented:
Done.
